File: Algorithms/p101_Symmetric_Tree_1.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return bool(not root or self.isSymmetricHelper(root.left,root.right))
        
    def isSymmetricHelper(self, t1, t2):
        return bool((not t1 and not t2) or (t1 and t2 and t1.val==t2.val and self.isSymmetricHelper(t1.left,t2.right) and self.isSymmetricHelper(t1.right,t2.left)))

File: Algorithms/p101_Symmetric_Tree/p101_Symmetric_Tree_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return !root || isSymmetricHelper(root->left,root->right);
    }
    bool isSymmetricHelper(TreeNode* t1, TreeNode* t2) {
        return !t1&&!t2 || t1&&t2&&t1->val==t2->val&&isSymmetricHelper(t1->left,t2->right)&&isSymmetricHelper(t1->right,t2->left);
    }
};
